c/secp256k1_ripemd160_sha256_sighash_all.c

```diff
@@ -0,0 +1,214 @@
+    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+  }
+
+  secp256k1_context context;
+  ret = ckb_secp256k1_custom_verify_only_initialize(&context, secp_data);
+  if (ret != 0) {
+    return ret;
+  }
+
+  len = BLAKE2B_BLOCK_SIZE;
+  ret = ckb_load_tx_hash(tx_hash, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+
+  while (1) {
+    len = 0;
+    /*
+     * Actually we don't need this syscall, we are just making it to grab all
+     * input indices for current group, from which we can load the actual
+     * witness data we need. `since` field is chosen here since it has a fixed
+     * size of 8 bytes, which is both predictable, and also provides minimal
+     * cycle consumption.
+     */
+    ret = ckb_load_input_by_field(NULL, &len, 0, index, CKB_SOURCE_GROUP_INPUT,
+                                  CKB_INPUT_FIELD_SINCE);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      return 0;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ERROR_SYSCALL;
+    }
+
+    /* Now we load actual witness data using the same input index above. */
+    len = WITNESS_SIZE;
+    ret = ckb_load_witness(witness, &len, 0, index, CKB_SOURCE_GROUP_INPUT);
+    if (ret != CKB_SUCCESS) {
+      return ERROR_SYSCALL;
+    }
+    if (len > WITNESS_SIZE) {
+      return ERROR_WITNESS_TOO_LONG;
+    }
+
+    if (!(witness_table = ns(Witness_as_root(witness)))) {
+      return ERROR_ENCODING;
+    }
+    args = ns(Witness_data(witness_table));
+    if (ns(Bytes_vec_len(args)) < 2) {
+      return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+    }
+
+    /* load pubkey and check pubkey hash */
+    len = TEMP_SIZE;
+    ret = extract_bytes(ns(Bytes_vec_at(args, 1)), temp, &len);
+    if (ret != CKB_SUCCESS) {
+      return ERROR_ENCODING;
+    }
+
+    /* parse pubkey */
+    secp256k1_pubkey pubkey;
+    if (secp256k1_ec_pubkey_parse(&context, &pubkey, temp, len) == 0) {
+      return ERROR_SECP_PARSE_PUBKEY;
+    }
+
+    sha256_state sha256_ctx;
+    ripemd160_state ripe160_ctx;
+
+    sha256_init(&sha256_ctx);
+    sha256_update(&sha256_ctx, temp, len);
+    sha256_finalize(&sha256_ctx, temp);
+
+    ripemd160_init(&ripe160_ctx);
+    ripemd160_update(&ripe160_ctx, temp, SHA256_SIZE);
+    ripemd160_finalize(&ripe160_ctx, temp);
+
+    /* check pubkey hash */
+    if (memcmp(argv[1], temp, RIPEMD160_SIZE) != 0) {
+      return ERROR_PUBKEY_RIPEMD160_HASH;
+    }
+
+    /* Load signature */
+    len = TEMP_SIZE;
+    ret = extract_bytes(ns(Bytes_vec_at(args, 0)), temp, &len);
+    if (ret != CKB_SUCCESS) {
+      return ERROR_ENCODING;
+    }
+
+    secp256k1_ecdsa_signature signature;
+    if (secp256k1_ecdsa_signature_parse_compact(&context, &signature, temp) == 0) {
+        return ERROR_SECP_PARSE_SIGNATURE;
+    }
+
+    /* verify signature */
+    blake2b_state blake2b_ctx;
+    blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+    blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
+    for (size_t i = 1; i < ns(Bytes_vec_len(args)); i++) {
```

Comment:
I thought at one point we want to remove the digest on witnesses?

Reply:
If we do not verify the signature, we will not be able to prove that the trader owns the private key corresponding to the public key.

Reply:
I mean you can only use tx_hash as secp message, you don't have to hash additional witness arguments, you can require witness only has one signature argument

Reply:
This special handling will result in inconsistency with the default lock behavior. Transaction sign requires special handling. Is it necessary to have such specialization?

Reply:
I think this lock has different use cases from default lock, it serves its own purpose and don't need those. Removing those ensures we can remove blake2b as well, which further decreases cell size.

Reply:
ok，done